cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i

```diff
@@ -437,49 +484,59 @@ template <typename algorithmFPType, CpuType cpu>
     ImpurityData right;
-    bool bFound = false;
     algorithmFPType vBest;
+    bool bFound               = false;
     size_t nDiffFeatureValues = 0;
+    auto aResponse            = this->_aResponse.get();
+    auto aWeights             = this->_aWeights.get();
+
     for (size_t i = 0; i < n - nMinSplitPart;)
     {
         ++nDiffFeatureValues;
-        size_t count                = 1;
-        const algorithmFPType first = featureVal[i];
-        const size_t iStart         = i;
-        for (++i; (i < n) && (featureVal[i] == first); ++count, ++i)
-            ;
-        if ((count < nMinSplitPart) || ((n - count) < nMinSplitPart) || (count < minWeightLeaf) || ((n - count) < minWeightLeaf)) continue;
+        size_t count                   = 1;
+        const algorithmFPType firstVal = featureVal[i];
+        const size_t iStart            = i;
+        algorithmFPType leftWeights    = aWeights[aIdx[i]].val;
+        for (++i; (i < n) && (featureVal[i] == firstVal); ++count, ++i)
+        {
+            leftWeights += aWeights[aIdx[i]].val;
+        }
+        if ((count < nMinSplitPart) || ((n - count) < nMinSplitPart) || (leftWeights < minWeightLeaf)
+            || ((totalWeights - leftWeights) < minWeightLeaf))
+            continue;
 
         if ((i == n) && (nDiffFeatureValues == 2) && bFound) break; //only 2 feature values, one possible split, already found
 
-        calcImpurity(aIdx + iStart, count, left);
-        subtractImpurity<algorithmFPType, cpu>(curImpurity.var, curImpurity.mean, left.var, left.mean, count, right.var, right.mean, n - count);
-#ifdef DEBUG_CHECK_IMPURITY
-        if (iStart == 0) checkImpurityInternal(aIdx + count, n - count, right);
-#endif
-        const algorithmFPType v = algorithmFPType(count) * left.var + algorithmFPType(n - count) * right.var;
+        algorithmFPType weights = 0.;
+        calcImpurity(aIdx + iStart, count, left, weights);
+        DAAL_ASSERT(fabs(weights - leftWeights) < 0.001);
+        subtractImpurity<algorithmFPType, cpu>(curImpurity.var, curImpurity.mean, left.var, left.mean, leftWeights, right.var, right.mean,
+                                               totalWeights - leftWeights);
+        const algorithmFPType v = leftWeights * left.var + (totalWeights - leftWeights) * right.var;
         if (!bFound || v < vBest)
         {
             vBest              = v;
             split.left.var     = left.var;
             split.left.mean    = left.mean;
             split.nLeft        = count;
+            split.leftWeights  = leftWeights;
             split.iStart       = iStart;
-            split.featureValue = first;
+            split.featureValue = firstVal;
             bFound             = true;
         }
     }
     if (bFound)
     {
-        algorithmFPType impurityDecrease = curImpurity.var - vBest / algorithmFPType(n);
+        algorithmFPType impurityDecrease = curImpurity.var - vBest / isPositive<algorithmFPType, cpu>(totalWeights) ? totalWeights : 1.0;
```

Comment:
Inconsistent const style for 1.0 (vs 1.)

const algorithmFPType impurityDecrease ?

Reply:
yes, fixed